### ISSUE TITLE
ORT Training: Training mnist using provided sample using a convolution model

### ISSUE THREAD

**Issue by georgen117** [microsoft/onnxruntime#4762]
I followed instructions posted by @liqunfu in #3706 to successfully run the provided mnist training sample.

> I attached the mnist test data and models. Please run with provided command line. As you can see, to work with cpp API, you need to follow the mnist example to build the graph with loss output so that backprop graph can be constructed.
>
> [mnist.zip](https://github.com/microsoft/onnxruntime/files/4553416/mnist.zip)
> ./onnxruntime_training_mnist --model_name ~/mnist/mnist_gemm_simple --train_data_dir ~/mnist/mnist_data/

However if the training model uses convolution (conv) instead of gemm I have not been able to successfully run training.

The download linked provided by @liqunfu that was posted in #3706 contains a training model with convolution `mnist_conv.onnx`

When trying that sample using the following command:

    ./onnxruntime_training_mnist --model_name ~/mnist/mnist_conv --train_data_dir ~/mnist/mnist_data/

when I run this I got the following error:

```
Loading MNIST data from folder ~\mnist\mnist_data\
Preparing data ...
Preparing data: done
#training set size = 60000
#test set size = 10000
2020-08-11 15:04:00.3371517 [W:onnxruntime:Default, graph.cc:846 onnxruntime::Graph::Graph] Initializer W1 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2020-08-11 15:04:00.3379735 [W:onnxruntime:Default, graph.cc:846 onnxruntime::Graph::Graph] Initializer W2 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2020-08-11 15:04:00.3386530 [W:onnxruntime:Default, graph.cc:846 onnxruntime::Graph::Graph] Initializer W3 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2020-08-11 15:04:00.3393177 [W:onnxruntime:Default, graph.cc:846 onnxruntime::Graph::Graph] Initializer B1 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2020-08-11 15:04:00.3465607 [W:onnxruntime:Default, graph.cc:846 onnxruntime::Graph::Graph] Initializer B2 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2020-08-11 15:04:00.3473163 [W:onnxruntime:Default, graph.cc:846 onnxruntime::Graph::Graph] Initializer B3 appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
2020-08-11 15:04:00.3482157 [W:onnxruntime:Default, graph.cc:846 onnxruntime::Graph::Graph] Initializer shape appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
Fail: This is an invalid model. Error in Node:Conv_0_Grad/ConvGrad_0 : Node (Conv_0_Grad/ConvGrad_0)'s output 0 is marked single but has an empty string in the graph
```

I created my own model using pytorch and exported that in `.onnx` format and tried to use that model instead.

I got almost the same failure:
```
Loading MNIST data from folder ~\mnist\mnist_data\
Preparing data ...
Preparing data: done
#training set size = 60000
#test set size = 10000
Fail: This is an invalid model. Error in Node:Conv_0_Grad/ConvGrad_0 : Node (Conv_0_Grad/ConvGrad_0)'s output 0 is marked single but has an empty string in the graph
```

The mnist sample has some comments in the code that seem to indicate it is designed to work with convolution.  Can anyone help me figure out what I am doing wrong when trying to run the minist training sample when the provided model contains a convolution operator.

Reference
#3706
#4249

EDIT:
Since this issue has been filed the onnxruntime_training_mnist sample has been updated so you can run a convolution model without modifying the source code by running the following command:

    ./onnxruntime_training_mnist --model_name ~/mnist/mnist_conv --train_data_dir ~/mnist/mnist_data/ --model_type conv

This was the command used when I was asked to check https://github.com/microsoft/onnxruntime/pull/7083

**Comment by hariharans29** [comment]
CC: @SherlockNoMad @manashgoswami for training related queries

**Comment by chethanpk** [comment]
I digged further into the issue and this is what i found out:
This is when it is trying to create a node for the conv gradient. With respect to the first conv operator and when it is trying to create its gradient node, it is failing a condition for one of its outputs ( IsGradientRequiredForSrcNodeInput ), and ends up creating a null ptr to the output which is failing one of the checks in the graph resolve function (as the output string associated with it is also null).

The place where the node is created:
onnxruntime\orttraining\orttraining\core\graph\gradient_builder.cc:
IMPLEMENT_GRADIENT_BUILDER(GetConvGradient) {
  std::vector<ArgDef> outputs;
  for (int i = 0; i < 3; i++) {
    if (IsGradientRequiredForSrcNodeInput(i)) {
    if (1) {
      outputs.push_back(GI(i));
    } else {
      outputs.push_back(ArgDef("", nullptr));
    }
  }

  return std::vector<NodeDef>{
      NodeDef("ConvGrad",
              {GO(0), I(0), I(1)},
              outputs,
              SrcNodeAttributes())};
}

And the check:
onnxruntime\orttraining\orttraining\core\graph\gradient_builder_base.h
 bool IsGradientRequiredForSrcNodeInput(const size_t i) const {
    return i < node_->InputDefs().size() &&
           gradient_outputs_.find(node_->InputDefs()[i]->Name()) != gradient_outputs_.end();
  }


Is this the correct behaviour?

**Comment by georgen117** [comment]
As @chethanpk pointed out the MNIST training sample can run using a convolution based model if the following changes are made to the code.

```
--- a/orttraining/orttraining/models/mnist/main.cc
+++ b/orttraining/orttraining/models/mnist/main.cc
@@ -30,7 +30,7 @@ using namespace onnxruntime::training::tensorboard;
using namespace std;

const static int NUM_CLASS = 10;
-const static vector<int64_t> IMAGE_DIMS = {784};  //{1, 28, 28} for mnist_conv
+const static vector<int64_t> IMAGE_DIMS = {1, 28, 28};  //{1, 28, 28} for mnist_conv
const static vector<int64_t> LABEL_DIMS = {10};


--- a/orttraining/orttraining/core/graph/gradient_builder.cc
+++ b/orttraining/orttraining/core/graph/gradient_builder.cc
@@ -705,7 +705,7 @@ IMPLEMENT_GRADIENT_BUILDER(GetTrainableDropoutGradient) {
IMPLEMENT_GRADIENT_BUILDER(GetConvGradient) {
   std::vector<ArgDef> outputs;
   for (int i = 0; i < 3; i++) {
-    if (IsGradientRequiredForSrcNodeInput(i)) {
+    if (/*IsGradientRequiredForSrcNodeInput(i)*/1) {
       outputs.push_back(GI(i));
     } else {
       outputs.push_back(ArgDef("", nullptr));
```

The `IsGradientRequiredForSrcNodeInput` check was causing the invalid model failure reported above.

```
Fail: This is an invalid model. Error in Node:Conv_0_Grad/ConvGrad_0 : Node (Conv_0_Grad/ConvGrad_0)'s output 0 is marked single but has an empty string in the graph
```

By bypassing the failure check the model could be run. However, I am unsure why the check is there and if the check should be required or updated.

**Comment by wschin** [comment]
I think this is a bug in ORT training. The definition of ConvGrad is
```c++
  ONNX_CONTRIB_OPERATOR_SCHEMA(ConvGrad)
      .SinceVersion(9)
      .Input(0, "dY", "Gradient of output Y", "T")
      .Input(1, "X", "Input tensor", "T")
      .Input(2, "W", "Weight tensor", "T")
      .Output(0, "dX", "Gradient of input X", "T")
      .Output(1, "dW", "Gradient of W", "T")
      .Output(2, "dB", "Gradient of B", "T")
      .AllowUncheckedAttributes()
      .TypeConstraint(
          "T",
          {"tensor(float16)", "tensor(float)", "tensor(double)"},
          "Constrain input and output types to float tensors.");
```
As you can see, there is no `Optional` flag (see RNN for the use of `Optional` flag). Because all outputs are not optional, we cannot use tensor name `""` (aka `ArgDef("", nullptr)`) in its output list.

**Comment by SherlockNoMad** [comment]
ConvGrad is a partially implemented op and I won't recommend using this op for your experiment. 
If your intention is to just try out ORT training with some simple model, I would recommend using a MNIST model with only Gemm or Matmul op. 
Alternatively, you can please try out https://github.com/microsoft/onnxruntime/blob/master/orttraining/pytorch_frontend_examples/mnist_training.py 

Accessing ORT through our pytorch frontend is the more of a recommended way.

**Comment by georgen117** [comment]
@SherlockNoMad thanks for the response. It may interest you to know once we applied the changes shown in the diff in the comment above I have been able to use the convolution model that was found in the download posted #3706 to train the C++ mnist sample to 95% accuracy.

I did have to run the remove_initializer_from_input.py script on the model.

Thanks for the pointer to https://github.com/microsoft/onnxruntime/blob/master/orttraining/pytorch_frontend_examples/mnist_training.py
I have already used that sample but not worrying about python bindings has been helpful when debugging this issue.

**Comment by georgen117** [comment]
This is still an issue. We have been manually applying the patch that was shown in my september 4 post so we can work with Convolution based models.

```
--- a/orttraining/orttraining/core/graph/gradient_builder.cc
+++ b/orttraining/orttraining/core/graph/gradient_builder.cc
@@ -705,7 +705,7 @@ IMPLEMENT_GRADIENT_BUILDER(GetTrainableDropoutGradient) {
IMPLEMENT_GRADIENT_BUILDER(GetConvGradient) {
   std::vector<ArgDef> outputs;
   for (int i = 0; i < 3; i++) {
-    if (IsGradientRequiredForSrcNodeInput(i)) {
+    if (/*IsGradientRequiredForSrcNodeInput(i)*/ true) {
       outputs.push_back(GI(i));
     } else {
       outputs.push_back(ArgDef("", nullptr));
```

**Comment by pranav-prakash** [comment]
@georgen117
`for (int i = 0; i < 3; i++) ` only works if the conv has the bias right? If there's no bias then i=2 is an invalid index. I'm pretty sure that's what the `IsGradientRequiredForSrcNodeInput` was trying to check
